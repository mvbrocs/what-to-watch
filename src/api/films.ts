import { appAxios } from './';

export interface IFilm {
  name: string;
  poster_image: string;
  preview_image: string;
  background_image: string;
  background_color: string;
  description: string;
  rating: number;
  scores_count: number;
  director: string;
  starring: string[];
  run_time: number;
  genre: string;
  released: number;
  id: number;
  is_favorite: boolean;
  video_link: string;
  preview_video_link: string;
}

export const filmsAPI = {
  getAll: () => appAxios.get<IFilm[]>('/films'),
};
