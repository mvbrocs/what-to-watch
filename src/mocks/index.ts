export { default as mockFilms } from './films';
export { default as mockStore } from './store';
