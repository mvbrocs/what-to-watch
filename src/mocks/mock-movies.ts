import img1 from './img/fantastic-beasts-the-crimes-of-grindelwald.jpg';
import img2 from './img/bohemian-rhapsody.jpg';
import img3 from './img/macbeth.jpg';
import img4 from './img/aviator.jpg';
import img5 from './img/we-need-to-talk-about-kevin.jpg';

export interface IMovie {
  id: string;
  title: string;
  genre: string;
  img: {
    src: string;
    alt: string;
    width: number;
    height: number;
  };
  videoSrc: string;
}

export const mockMovies: IMovie[] = [
  {
    id: 'movie-1',
    title: 'Fantastic Beasts: The Crimes of Grindelwald',
    genre: 'Fantastic',
    img: {
      src: img1,
      alt: 'Fantastic Beasts: The Crimes of Grindelwald',
      width: 280,
      height: 175,
    },
    videoSrc:
      'https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4',
  },
  {
    id: 'movie-2',
    title: 'Bohemian Rhapsody',
    genre: 'Dramas',
    img: {
      src: img2,
      alt: 'Bohemian Rhapsody',
      width: 280,
      height: 175,
    },
    videoSrc:
      'https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm',
  },
  {
    id: 'movie-3',
    title: 'Macbeth',
    genre: 'Comedy',
    img: {
      src: img3,
      alt: 'Macbeth',
      width: 280,
      height: 175,
    },
    videoSrc:
      'https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4',
  },
  {
    id: 'movie-4',
    title: 'Aviator',
    genre: 'Horror',
    img: {
      src: img4,
      alt: 'Aviator',
      width: 280,
      height: 175,
    },
    videoSrc:
      'https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm',
  },
  {
    id: 'movie-5',
    title: 'What We Do in the Shadows',
    genre: 'Crime',
    img: {
      src: img5,
      alt: 'What We Do in the Shadows',
      width: 280,
      height: 175,
    },
    videoSrc:
      'https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4',
  },
];
