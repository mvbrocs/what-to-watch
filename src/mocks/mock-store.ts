import { RootState } from 'state/root-reducer';

export const mockStore: RootState = {
  genre: 'Some genre',
  movies: {
    loaded: false,
    loading: false,
    error: null,
    maxVisible: 0,
    ids: [],
    entities: {},
  },
};
