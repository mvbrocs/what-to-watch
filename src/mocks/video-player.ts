import img1 from './img/fantastic-beasts-the-crimes-of-grindelwald.jpg';

export const videoPlayerAssetsMock = {
  src:
    'https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4',
  poster: {
    src: img1,
    alt: 'Fantastic Beasts: The Crimes of Grindelwald',
    width: 280,
    height: 175,
  },
};
