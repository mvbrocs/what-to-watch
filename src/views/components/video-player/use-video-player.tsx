import { useState, useEffect } from 'react';

type UseVideoPlayerArgs = {
  play: boolean;
  delay?: number;
  videoElement: null | HTMLVideoElement;
};

export const useVideoPlayer = ({ play, delay, videoElement }: UseVideoPlayerArgs) => {
  const [start, setStart] = useState(false);

  useEffect(() => {
    if (delay === undefined) {
      setStart(play);
      return;
    }

    const timeoutID = setTimeout(() => setStart(true), delay);

    if (!play) {
      setStart(false);
      clearTimeout(timeoutID);
    }

    return () => clearTimeout(timeoutID);
  }, [play, delay]);

  useEffect(() => {
    if (start && videoElement) {
      videoElement.currentTime = 0;
      videoElement.play();
    } else if (videoElement) {
      videoElement.pause();
    }
  }, [start, videoElement]);

  return { start };
};
