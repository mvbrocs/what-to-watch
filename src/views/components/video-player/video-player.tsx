import React, { useEffect, useRef } from 'react';

type VideoPlayerProps = {
  play: boolean;
  src: string;
};

export const VideoPlayer = ({ play, src }: VideoPlayerProps) => {
  const videoRef = useRef<HTMLVideoElement>(null);

  useEffect(() => {
    const videoElement = videoRef?.current;

    if (play && videoElement) {
      videoElement.play();
    } else if (videoElement) {
      videoElement.pause();
      videoElement.currentTime = 0;
    }
  }, [play]);

  return <video ref={videoRef} src={src} muted />;
};
