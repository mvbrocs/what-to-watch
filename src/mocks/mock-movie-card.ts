import { mockMovies } from './mock-movies';

export const mockMovieCard = mockMovies[0];
