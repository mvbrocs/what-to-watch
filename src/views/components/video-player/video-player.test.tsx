import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import { VideoPlayer } from 'views/components/video-player/video-player';
import { mockFilms } from 'mocks/films';

describe('VideoPlayer', () => {
  const assetsProps = {
    src: mockFilms[0].preview_video_link,
    poster: {
      src: mockFilms[0].preview_image,
      alt: mockFilms[0].name,
      width: 280,
      height: 175,
    },
  };
  describe('not play', () => {
    const videoPlayer = mount(<VideoPlayer {...assetsProps} play={false} />);

    it('renders properly', () => {
      expect(videoPlayer).toMatchSnapshot();
    });
  });

  describe('play', () => {
    describe('without delay', () => {
      const videoPlayer = mount(<VideoPlayer {...assetsProps} play={true} />);

      it('renders properly', () => {
        expect(videoPlayer).toMatchSnapshot();
      });
    });

    describe('with delay', () => {
      jest.useFakeTimers();

      const videoPlayer = mount(<VideoPlayer {...assetsProps} play={true} delay={1000} />);

      it('renders properly', () => {
        expect(videoPlayer).toMatchSnapshot();
      });

      it('should render preview image when the delay not expired', () => {
        act(() => {
          jest.advanceTimersByTime(500);
        });
        videoPlayer.update();

        expect(videoPlayer).toMatchSnapshot();
      });

      it('should not render preview image when the delay expired', () => {
        act(() => {
          jest.advanceTimersByTime(500);
        });
        videoPlayer.update();

        expect(videoPlayer).toMatchSnapshot();
      });
    });
  });
});
