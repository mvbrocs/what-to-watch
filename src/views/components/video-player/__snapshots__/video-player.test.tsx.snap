// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer not play renders properly 1`] = `
<VideoPlayer
  play={false}
  poster={
    Object {
      "alt": "Fantastic Beasts: The Crimes of Grindelwald",
      "height": 175,
      "src": "fantastic-beasts-the-crimes-of-grindelwald.jpg",
      "width": 280,
    }
  }
  src="https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4"
>
  <img
    alt="Fantastic Beasts: The Crimes of Grindelwald"
    height={175}
    src="fantastic-beasts-the-crimes-of-grindelwald.jpg"
    width={280}
  />
  <video
    height={175}
    muted={true}
    src="https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4"
    width={280}
  />
</VideoPlayer>
`;

exports[`VideoPlayer play with delay should not render preview image when the delay expired 1`] = `
<VideoPlayer
  delay={1000}
  play={true}
  poster={
    Object {
      "alt": "Fantastic Beasts: The Crimes of Grindelwald",
      "height": 175,
      "src": "fantastic-beasts-the-crimes-of-grindelwald.jpg",
      "width": 280,
    }
  }
  src="https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4"
>
  <video
    height={175}
    muted={true}
    src="https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4"
    width={280}
  />
</VideoPlayer>
`;

exports[`VideoPlayer play with delay should render preview image when the delay not expired 1`] = `
<VideoPlayer
  delay={1000}
  play={true}
  poster={
    Object {
      "alt": "Fantastic Beasts: The Crimes of Grindelwald",
      "height": 175,
      "src": "fantastic-beasts-the-crimes-of-grindelwald.jpg",
      "width": 280,
    }
  }
  src="https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4"
>
  <img
    alt="Fantastic Beasts: The Crimes of Grindelwald"
    height={175}
    src="fantastic-beasts-the-crimes-of-grindelwald.jpg"
    width={280}
  />
  <video
    height={175}
    muted={true}
    src="https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4"
    width={280}
  />
</VideoPlayer>
`;

exports[`VideoPlayer play without delay renders properly 1`] = `
<VideoPlayer
  play={true}
  poster={
    Object {
      "alt": "Fantastic Beasts: The Crimes of Grindelwald",
      "height": 175,
      "src": "fantastic-beasts-the-crimes-of-grindelwald.jpg",
      "width": 280,
    }
  }
  src="https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4"
>
  <video
    height={175}
    muted={true}
    src="https://st.depositphotos.com/1100827/3666/v/600/depositphotos_36660843-stock-video-girl-leans-back-among-the.mp4"
    width={280}
  />
</VideoPlayer>
`;
